src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs

```diff
@@ -54,35 +56,77 @@ public class CollectionQueryEnumerable : IEnumerable<Document>
             _includeCompareExchangeValuesCommand = includeCompareExchangeValuesCommand;
             _totalResults = totalResults;
             _token = token;
+            _scannedResults = scannedResults;
+            _skippedResults = skippedResults;
         }
 
         public string StartAfterId { get; set; }
 
         public Reference<long> AlreadySeenIdsCount { get; set; }
 
-        public Reference<long> SkippedResults { get; set; }
-
         public DocumentFields Fields { get; set; } = DocumentFields.All;
 
         public IEnumerator<Document> GetEnumerator()
         {
-            return new Enumerator(_database, _documents, _fieldsToFetch, _collection, _isAllDocsCollection, _query, _queryTimings, _context, _includeDocumentsCommand, _includeRevisionsCommand, _includeCompareExchangeValuesCommand, _totalResults, StartAfterId, AlreadySeenIdsCount, Fields, SkippedResults, _token);
+            return new Enumerator(_database, _documents, _fieldsToFetch, _collection, _isAllDocsCollection, _query,
+                _queryTimings, _context, _includeDocumentsCommand, _includeRevisionsCommand, _includeCompareExchangeValuesCommand, _totalResults, _scannedResults, 
+                StartAfterId, AlreadySeenIdsCount, Fields, _skippedResults, _token);
         }
 
         IEnumerator IEnumerable.GetEnumerator()
         {
             return GetEnumerator();
         }
 
+        public class FilterKey : ScriptRunnerCache.Key
+        {
+            private readonly QueryMetadata _queryMetadata;
+
+            public FilterKey(QueryMetadata queryMetadata)
+            {
+                _queryMetadata = queryMetadata;
+            }
+
+            public override void GenerateScript(ScriptRunner runner)
+            {
+                if (_queryMetadata.DeclaredFunctions != null)
+                {
+                    foreach (var function in _queryMetadata.DeclaredFunctions)
+                    {
+                        runner.AddScript(function.Value.FunctionText);
+                    }
+                }
+
+                runner.AddScript(_queryMetadata.FilterScript);
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is FilterKey fk)
+                {
+                    return _queryMetadata == fk._queryMetadata;
```

Comment:
@ayende Just to make sure - this is checking the quality by reference. It's okay because we cache the metadata and we rely on that fact here, right?

Reply:
Yes, that is the idea.

Reply:
May be better to use `ReferenceEquals` directly.

Reply:
Changed.